### ISSUE TITLE
How do I tell FreshRSS to sort articles by publication date, not date added?

### ISSUE THREAD

**Issue by deanishe** [FreshRSS/FreshRSS#2596]
Every time I add new subscriptions, "Today" is filled with articles that are weeks or months old. And those new articles are in a nonsensical order (presumably some implementation detail):

![image](https://user-images.githubusercontent.com/747913/67589636-28319980-f759-11e9-9098-79918e6a8a55.png)

I would like to be able to mark all old items as read with one click, the way I can with existing subscriptions.

Thanks.

**Comment by Alkarex** [comment]
Hello,
You can search/filter by publication date. E.g. everything older than the first of october: `pubdate:/2019-10-01`

**Comment by deanishe** [comment]
Thanks. That helps.

So there's no way to set FreshRSS to actually sort by publication date?

**Comment by Alkarex** [comment]
@deanishe No, it is not currently possible to sort by publication date (by design, see several related issues on the topic that explain why). However, I am considering adding some new order mechanisms (e.g. by relevance / priority), during which it might nevertheless become possible to sort by publication date in the Web UI

**Comment by deanishe** [comment]
> by design, see several related issues on the topic that explain why

I've read the other issues, and I understand why that sort order works for you and your subscriptions. Unfortunately, it is poorly suited to me and mine.

> However, I am considering adding some new order mechanisms (e.g. by relevance / priority)

Then I'll keep my fingers crossed that you also add support for plain, old publication order.

Thanks for your help.

**Comment by Redsandro** [comment]
Hello,

I would like to indicate my interest in this as well.

I'm trying out RSS readers for the first time. FreshRSS looked the most appealing to me on screenshots. I start a FreshRSS docker container and add three feeds. This is literally the next thing I was trying to accomplish. I want to read my selected news in the order it was published. 

But it is ordered by feed first and by order in the XML file second or something? I don't believe it's very intuitive. The publish date is right there, on the right. I rub my eyes to see if I missed a little sorting triangle above that column. I reckon you can just click on the feed in the sidebar if you want to see only the items from a single feed without other feeds mixed in.

It is not a small preference; it was in fact wat I was trying to accomplish that made me try an RSS reader in the first place. Apart from that, FreshRSS is giving a good first impression. I like that I can self-host it and use the web-app on multiple devices. And it was super easy and fast to configure.

Please add some more weight to this feature request.

**Comment by Redsandro** [comment]
The more often you fetch feeds, the less of a problem this becomes. But certain feeds are more likely to block you from fetching too often.

**Comment by CoryMnvl** [comment]
I have mostly enjoyed setting up and using FreshRSS, thank you for all the time and effort!

I would like to add my voice to this feature request. This was one of my first annoyances when I started using FreshRSS and continues to be one. 

In theory I agree that the more feeds you have, and the less feeds you add, the less of a problem this is. However this morning I got excited seeing there was an update from the FreshRSS Github feed at the top of the list, which was added to my reader yesterday (27 February 2023), just to realize later that it was published 2020-05-31. 

This sorting behavior is misleading:
![image](https://user-images.githubusercontent.com/27926827/221910415-8229a90c-865a-48c9-8d55-5d7199ce8862.png)

**Comment by Frenzie** [comment]
What you're showing is something a bit different. One or both of these must be enabled (Settings → Archiving → Purge policy). I'd set them a lot higher or disable them completely, but ymmv and all that. The problem on your screenshot would occur regardless of sorting  due to the unread status.

![image](https://user-images.githubusercontent.com/202757/221933163-b7ccbd22-d366-47c1-9431-82e0cf12f449.png)

**Comment by CoryMnvl** [comment]
What I'm showing is an old article (nearly 3 years old), outside of my control, appearing recently at the top of the list as if it was new, just because FreshRSS recently became aware of it. It was meant to show a situation I came across that exacerbates this issue. 

My install of FreshRSS is less than 1 month old, and my purge policies did not save me from this (I assume they also operate on the imported instead of published date/time):
![image](https://user-images.githubusercontent.com/27926827/221944960-aef46333-8456-49e2-9155-51e03caee8b3.png)

**Comment by schobernoise** [comment]
I would also like to voice my interest. Since this is all free software I am VERY grateful for the work the developers put in.
A plain old "sort by" button on the top of the column would be great... how much work would it be to implement? Can I assist somehow?
Because as of now it is kinda hard to get ahold of all the mixed up articles (I get in about 150 every day and add about 1 Source a week).

**Comment by Alkarex** [comment]
@schobernoise Along the path of allowing several sorting orders, some next steps are:
1. Update the front-end for marking all the articles as read (and marking the selection as unread). Currently, we use a range based on article ID (interval with minimum and maximum), and that needs to be changed to pass an array of article IDs (so that it works which whichever sorting order). The back-end is already ready.
2. Address paging issues (probably enough to give a constraint based on max entry ID) to avoid problems when loading more articles
3. Update the SQL query returning the articles, to accept different sorting orders
4. New UI to change sorting order

**Comment by kaysond** [comment]
+1. This is also a very weird behavior for when you first install FreshRSS and get everything set up. You get chunks of articles based on the order in which you added the feeds, rather than in publication order.

**Comment by djnavarro** [comment]
Also chiming in as a +1, as a new FreshRSS user who really loves it! There are definitely some situations where it's nice to be able to sort by the order in which posts were added (e.g. when I discover a new author and want to browse through all their old posts), but most of the time my reading behaviour involves "browsing backwards in time"), and it's a little jarring not to have the option of sorting by publication date.

**Comment by lonseidman** [comment]
Let me add another +1 to this. I am using FreshRSS as a secondary reader for an upcoming project and this sortation stategy is an awful way to introduce a new user to RSS. Every time I add a new feed it takes over the category and pushes everything else out. I get that you want to add the newest added item to the top of the list each time irrespective of declared pub date but this isn't working for me. 

Can anyone recommend a database command I could use to re-sort after importation?

**Comment by alexkutsan** [comment]
Same for me, this annoying bug/feature significantly breaks the UX (user experience) for newcomers and deters from using this awesome product. When I add a new source, it just pollutes theincome feed with very old posts in it.

**Comment by nickian** [comment]
This is a major flaw of FreshRSS, for me. It's not only a problem that adding a new source ends up at the top; it's also that when viewing a category, the articles should not be clumped by the feed source. They should be by date/time. It's obviously possible to fix, since this isn't a problem for Reeder or other native apps that are able to sync with the FreshRSS API.

**Comment by arkhi** [comment]
Hello,

I have also been looking to sort by published date instead of fetched date because this is how I’m used to browse articles.

That being said, I see the interest in having the clutter visible right away. It is probably a good thing for some users in some cases, just not the most common UX out there as far as I can tell.

So… One more thumb up for an option to sort as one sees fit. :)

**Comment by trebory6** [comment]
+1 for me too. 

Let's all be real, this is a self hosted list based aggregator of RSS feeds. One of the main features of almost every RSS reader in history has been the ability to sort your feed. 

The fact that FreshRSS is an incredible oversight, one that is defended by "The more you use it the less of an issue it is," which isn't a good way to justify anything really. 

The fact I can just select my feed and paste it into a spreadsheet and sort it from there is proof alone that it's relatively simple to add this functionality. If FreshRSS was in Python I'd be able to convert the HTML of the main feed to a CSV and create a UI to display and sort it with ease, In fact I could probably write a browser plugin that does this. 

Anyways, I might actually work on it and inspect the code of FreshRSS to see what you guys are talking about it being so difficult, but I just wanted to push back against a lot of these excuses.  Really really ridiculous this isn't a basic feature.

**Comment by lonseidman** [comment]
Good to see continued advocacy for a sortation feature. Maybe they could implement a maintenance task that will reorder things? I am coming back to this thread having tested every self-hosted browser-based RSS reader I could find and there is nothing better than FreshRSS.. If this issue was resolved it would by far be the best solution.

**Comment by Redsandro** [comment]
> an incredible oversight
> which isn't a good way to justify anything
> it's relatively simple to add

Hi, I've advocated for this feature myself somewhere above. I'd like to remind everyone that this is a free and open source project made by a collaboration of people like you. If you find certain features lacking, you can politely indicate interest. If no one wants to pick up the job, the best you can do is implement it yourself. Something relatively simple is a good first commit. You shouldn't really get worked up over something free, right?

**Comment by CarpetbaggerFL** [comment]
Another +1 for adding this important feature. It may not mean a lot if you're fetching blog entries but for following news stories as they develop it's essential.

**Comment by MonkWho** [comment]
I'd like to add my upvote to this as well. I just imported a bunch of my feeds from Feedly to FreshRSS and now all historic articles are sorted by today's date instead of publication date. I understand that with a cron job in place all new articles going forward will be OK'ish but would love to be able to have proper sorting for older articles as well without having to apply custom code.

**Comment by Lcchy** [comment]
I would very much like to have this feature too so I have started to look at the code but as I have no experience with Freshrss, if anyone wants/can help, that would be amazing!

(The fix seems to work but the scrolling pages are not behaving well)

This is the PR #5776

**Comment by X1Aaron** [comment]
+1

**Comment by MathieuGodard** [comment]
+1

**Comment by lordmatt** [comment]
+1

I would love to sort by publication date

**Comment by jsspen** [comment]
After setting up FreshRSS for a friend today I was reminded of how horrible it is to not have this functionality and how much it dissuades new users from using this otherwise great app. @Alkarex can we please get an update on this functionality, it's been nearly a year since the last one...

**Comment by olivier-duval** [comment]
+1

Actually, sort by publication date is the only logical possible sort.

**Comment by s1lverkin** [comment]
Yes, I would love to have that feature

**Comment by leapmotiondev** [comment]
yep, i just installed freshrss again as a news screener for my stocks and this feature would be highly appreciated!
i might have some spare time to create an extension if this keeps getting requested.

**Comment by kupietools** [comment]
Yes, as a heavy RSS user, the weird, noninuitive sort order is the one thing that keeps me looking for something besides freshRSS, it’s basically a dealkiller in what would otherwise be a just about perfect solution (and, yes, one which I greatly appreciate the volunteer effort that has gone into developing.) At some point, if I have time, I’ll look at the code myself… I don’t even know what this is written in, if it’s PHP, I may be able to implement this myself; if python probably not. Either way I’m sure it’ll be a very long time before I have the kind of time on my hands to get involved with a whole new one completely unfamiliar project, it’s just kind of a pipe dream and I feel like my time would be better spent looking for another RSS reader than trying to familiarize myself with us enough to code what I need myself. So, while I recognize that nobody has any obligation to implement a feature that I need, it would be really great if someone already familiar with the code did, and I’d be grateful. Especially considering, if you look at this thread, how many years this has been requested for.

**Comment by Orlblr** [comment]
I concur with the previous commentators, I think this is the most important thing in a RSS reading tool

**Comment by flashlab** [comment]
Since almost 5 year past, Was this feature still in plan or anyone have compromised solution?

**Comment by Alkarex** [comment]
Guys, please stop commenting without having done your homework of reading and understanding the past comments, highlighted answers https://github.com/FreshRSS/FreshRSS/discussions/5865, and without proposing pull requests [addressing the plan](https://github.com/FreshRSS/FreshRSS/issues/2596#issuecomment-1541934281).
You can just vote on the top comment or a more precise one below.

**Comment by flashlab** [comment]
@Alkarex Thanks for the reminding. I really appriciate contributions of FreshRSS community members. What I intrest in is whether anyone presist in FreshRSS who can't do without the sorting that made some temporary workaround. It will be a great meaning to me. Found a hard code solution here -> #2083

**Comment by colin969** [comment]
It's understandable that people are going to come back to this problem when it's an immediate part of the user experience and it's been unsolved for 5 years.

The one PR that did crop up being immediately closed, and linked back to a comment without much discussion is also rather discourging. There isn't much point trying to put these things forward when evidently the core contributors aren't interested in it.

**Comment by Frenzie** [comment]
It seems important to point out that the comment in question is https://github.com/FreshRSS/FreshRSS/issues/2596#issuecomment-1541934281

**Comment by jsspen** [comment]
> Guys, please stop commenting without having done your homework of reading and understanding the past comments, highlighted answers #5865, and without proposing pull requests [addressing the plan](https://github.com/FreshRSS/FreshRSS/issues/2596#issuecomment-1541934281). You can just vote on the top comment or a more precise one below.

This all seems a bit circular. You are linking to your comment _from over a year ago_ that laid out a loose "plan" that you haven't given any updates on since (I tagged you and asked if you could provide an update in this thread a few months ago) and a link to an "answer" you gave in a Q&A thread 6 months ago about this problem where you explained that this behavior is intentional before linking to... this thread. I think it makes sense that people are congregating here to express their desire to have this and it seems odd to tell them _not_ to comment here...

**Comment by Alkarex** [comment]
@jsspen Let me develop a bit (even though [others have kindly tried already](https://github.com/FreshRSS/FreshRSS/issues/2596#issuecomment-1680821033)). This is mostly a question of priorities. There are plenty of other issues and also some good open PRs that have been waiting a long time. A bunch of improvements have materialised recently. Being an open source project developed on peoples' free time, the most efficient way to get a feature you want is to code it yourself, or to motivate someone to do it. As a matter of fact, there has not yet been any contribution helping with a PR to address even the first point of [what needs to be done](https://github.com/FreshRSS/FreshRSS/issues/2596#issuecomment-1541934281).
And personally, I do not get motivated by negative comments of this type, quite the contrary, and the energy spent here further delays the process.
What does motivate me in particular is people displaying a good understanding of the current behaviour as well as what other systems are doing (many systems including e-mail clients, SMS, and other message-based systems are doing the same than the current behaviour, as described before), and come with a precise descriptions of use-cases for which this alternative sorting option is essential (rather than a cosmetic preference). Especially when coming from people who have already helped the project in one way or another.
I am genuinely interested in getting more sorting options (including alphabetic, random, by similarity...) so even if let to me, this will be done (the issue is actually tentatively scheduled for 1.25.0). But it does require some work and is not something that can be rushed in one hour: there are many consequences to consider, including adding some DB indexes, work on performance, incompatibilities with the API, rework the UI, make new options, translations, tests, and not the least some documentation to mitigate the fact that this sorting mode is unstable.

**Comment by jsspen** [comment]
@Alkarex to be clear my comment was not in any way meant as criticism and I'm sorry that you seem to have read it as a "negative" comment. I'm very appreciative of the work you and the other contributors have put into this project. I do however think the conversation around this issue has been a bit confusing and a long absence of communication didn't help.

Personally I felt there hadn't (until now) been any real clarity on whether or not this was something that was being worked on by the core team and the material you suggested everyone here needed to understand before commenting didn't seem to make things any more clear, at least to me. My response above wasn't about criticizing or attempting to motivate anyone, it was simply looking for the actual status of this issue/feature. You've given that now and I appreciate that.

**Comment by Eoin-ONeill-Yokai** [comment]
Even if sorting by publication date isn't an option, it would perhaps be nice to at least "quietly" add RSS feeds and only track the new content when adding a feed. This could basically be done by immediately archiving or purging articles that are added from a new rss feed during the first importing.

I think most of our issues with the current implementation is that all content from a publication is pushed en-masse to the article list on addition.  At best this creates a lack of diversity of content when a new source is added but a more-or-less topical list of articles that are relevant for the category. At worst, it creates a wave of  news or articles from years ago masquerading as "new" content (especially on low traction blogs, such as personal blogs or niche subculture blogs.) For those who know, such as system admins, it's not a huge deal, but it can be confusing if you have guests or custom users who might not know the intricacies of the design. 

So if we could simply hide some of this content on the first wave of additions, that would at least help alleviate the problem without adding too many sorting features.

**Comment by Alkarex** [comment]
@Eoin-ONeill-Yokai While waiting for something better, the current way to automatically mark as read new articles that have a too old publication date is to add a general rule *Mark an article as read* with `!pubdate:P3d` adapting the duration (here, 3 days) for what you prefer.
See point 6 of https://github.com/FreshRSS/FreshRSS/discussions/6101#discussioncomment-8479032

**Comment by Eoin-ONeill-Yokai** [comment]
@Alkarex Awesome, thanks a lot for the help!

**Comment by chucksplatt** [comment]
Hello!
I recently started using FreshRSS and would like to see this feature/improvement get over the line. I am a developer familiar with PHP (it's been a few years, been mostly in Java/Python world recently). Is there an outstanding feature branch or an existing PR for this? I would like to take a crack at getting this done.

**Comment by Alkarex** [comment]
@chucksplatt That would be welcome. Ideally one PR per item of https://github.com/FreshRSS/FreshRSS/issues/2596#issuecomment-1541934281

**Comment by alecmaly** [comment]
I've made some [very] sloppy changes to my fork that \*seems\* to resolve pagination and marking as read. I'm not opening a pull request because I have only tested with a SQLite data source and I see more places in the code that may need to be changed for other use cases/integration, code quality could be improved, and the methodology for achieving this functionality may be a bit too messy for the repo maintainer's liking. 

https://github.com/alecmaly/FreshRSS/commit/9ff911ec69883590e7f4208f77a8bbc2e034bbc0

To test locally: 
```
git clone https://github.com/alecmaly/FreshRSS
cd FreshRSS && git checkout sort_by_publish
make start

# open browser to localhost:8080
```

Option to sort by date is in the "Reading" options menu, similar to [this video](https://youtu.be/rYo8u1syE70?si=6pNDfaW9B5vUN-1M&t=656) which I used as a starting point for changes. 

This is what I plan to use for my personal instance. Hopefully, this provides additional insight for someone to pick up and carry this across the finish line, I do not have the time to devote to a production ready solution. 

Happy holidays!

-----------------------

Additional notes:
- FreshRSS_Context::$offset has been repurposed, I saw it used in ./FreshRSS/app/views/helpers/htmlPagination.phtml but did not see this file referenced anywhere. Hopefully this does not break anything
- The string '000000' seems to be appended to dates (for WHERE clauses against the id field) and removing this suffex allows sql queries to properly retrieve articles by date. I localized changes to the functions that seemed to make my client work (e.g. [here](https://github.com/alecmaly/FreshRSS/commit/9ff911ec69883590e7f4208f77a8bbc2e034bbc0#diff-c9b33efaa35e9f9cb4b64ff8b8a97197b07fa6b78dd0a0d7166246916de19a42R494-R496)), but there are other functions (markReadCat(), markReadFeed(), markReadTag()) where I did not modify the $idMax value, I am not sure where these functions are used but they may need to be updated as well if continuing with this solution. 
- Additionally, when removing the '000000' I [check for a magic value of 12](https://github.com/alecmaly/FreshRSS/commit/9ff911ec69883590e7f4208f77a8bbc2e034bbc0#diff-c9b33efaa35e9f9cb4b64ff8b8a97197b07fa6b78dd0a0d7166246916de19a42R494) because timestamps from 1970-04-26 are 8 chars long, so when the 6 '0's are added this makes 12 - I believe this gives a decent range for publication dates. I want to prevent double removing of the appended '000000' in the event my removal code gets executed multiple times (I'm not 100% familiar with the callstacks/control flow) - this solution should work for thousands of years (until new timestamps are 12 chars in length). This was the lowest cost of effort for me; I would think a better solution could be to clean up all the places where '000000' is appended in the first place based on sort settings, but that may result in a lot of code changes and my solution should work for the foreseeable future (I think?).

**Comment by Alkarex** [comment]
Quick note: this is not forgotten, and I hope to find time soon. This one needs quite some care to be done properly. If some people here are willing to contribute, it would also be a great help to address some of the [simpler issues for which help is wanted](https://github.com/FreshRSS/FreshRSS/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22), to free some resources for this issue

**Comment by Alkarex** [comment]
Work in progress. Stay tuned 🎄

**Comment by Alkarex** [comment]
Tests welcome of this work-in-progress https://github.com/FreshRSS/FreshRSS/pull/7149 in particular for bugs / regressions.

The UI is not done yet, but the features can be experimented with:
* `&sort=id` (current behaviour, sorting according to newest received articles)
* `&sort=date` (publication date, which is not indicative of how new an article is)
* `&sort=title`
* `&sort=link`
* `&sort=rand` (random order - which disables infinite scrolling, at least for now)

combined with `&order=ASC` or `&order=DESC`

**Comment by Alkarex** [comment]
@alecmaly I have written some implementation notes in https://github.com/FreshRSS/FreshRSS/pull/7149 to explain why your initial approach based on `OFFSET` could not work

Tests still welcome
